MetaLamp/lending-pool/src/Plutus/Contracts/LendingPool/OnChain/Core/Validator.hs

```diff
@@ -166,178 +153,123 @@ validateWithdraw aave ReserveFundsDatum ctx (reserveId, actor) =
+  toBoolPrefixed "validateBorrow: " $ do
+    assertValidCurrentSlot ctx slot
+    newUserConfigs <- findUserConfigs ctx state
+    assertInsertAt userConfigId userConfigs newUserConfigs
+
+    let txInfo = scriptContextTxInfo ctx
+    oracleValues <-
       case foldrM (\o@(_, _, _, oAsset) acc -> fmap ((: acc) . (oAsset, )) (Oracle.findOracleValueInTxInputs txInfo o)) [] oracles of
-        Just vs -> AssocMap.fromList vs
-        _       -> traceError "validateBorrow: Oracles have not been provided"
-
-    isValidUserConfigsTransformation :: Bool
-    isValidUserConfigsTransformation =
-      maybe False checkUserConfigs userConfigsOutputDatum
-    checkUserConfigs ::
-         (AssetClass, AssocMap.Map UserConfigId UserConfig) -> Bool
-    checkUserConfigs (newStateToken, newUserConfigs) =
-      newStateToken == stateToken && doesCollateralCoverDebt actor oracleValues newUserConfigs &&
-      maybe False (checkRedeemerConfig $ AssocMap.lookup userConfigId userConfigs) (AssocMap.lookup userConfigId newUserConfigs)
-    checkRedeemerConfig :: Maybe UserConfig -> UserConfig -> Bool
-    checkRedeemerConfig oldState newState =
-      let debtAmount = (ucDebt newState -) $ maybe 0 ucDebt oldState
-          disbursementAmount = assetClassValueOf actorRemainderValue reserveId - assetClassValueOf actorSpentValue reserveId
-       in debtAmount == disbursementAmount && debtAmount > 0 && disbursementAmount > 0 &&
-          ucCollateralizedInvestment newState == 0 && maybe True ((== 0) . ucCollateralizedInvestment) oldState
-
-validateBorrow aave (ReservesDatum stateToken reserves) ctx userConfigId oracles =
+        Just vs -> pure . AssocMap.fromList $ vs
+        _       -> throwError "Oracles have not been provided"
+    reserve <- findReserves ctx state >>= maybe (throwError "Reserve not found") pure . AssocMap.lookup reserveId
+    unless (doesCollateralCoverDebt actor oracleValues newUserConfigs) $ throwError "Not enough collateral"
+
+    let oldState = AssocMap.lookup userConfigId userConfigs
+    newState <- maybe (throwError "User config not found") pure $ AssocMap.lookup userConfigId newUserConfigs
+    let accState = fmap (updateConfigAmounts reserve slot) oldState
+        actorSpentValue = valueSpentFrom txInfo actor
+        actorRemainderValue = valuePaidTo txInfo actor
+        debtAmount = ((iaAmount . ucDebt) newState -) $ maybe (fromInteger 0) (iaAmount . ucDebt) accState
+        disbursementAmount = fromInteger $ assetClassValueOf actorRemainderValue reserveId - assetClassValueOf actorSpentValue reserveId
+    unless
+      (debtAmount == disbursementAmount && debtAmount > fromInteger 0 && disbursementAmount > fromInteger 0 &&
+          (iaAmount . ucCollateralizedInvestment $ newState) == (fromInteger 0) && maybe True ((== (fromInteger 0)) . iaAmount . ucCollateralizedInvestment) oldState)
+      (throwError "")
```

Comment:
Let's fill this with meaningful error messages (with some generic message like "borrow reserves change not valid" for now)